### PR TITLE
Don't use output hash in filename for crates of type cdylib

### PR DESCRIPTION
**Pull request by ddeville**
Cdylib crates, similarly to binaries, are meant for consumption outside of the rust ecosystem. For this reason, there is no need to attempt making their name unique to disambiguate them from other crates with the same name but different features/versions in the Rust/Cargo dependency graph (as discussed in https://github.com/bazelbuild/rules_rust/issues/405).

In fact, when distributing dynamic libraries or embedding them into applications, it is *critical* for their name to be deterministic since it ends up embedded into the executable itself, that the linker uses to link/resolve the library. On macOS, this requires the use of `install_name_tool` to tweak the id/name after it's build. On Windows, one has to re-generate the LIB file to embed the new name (which afaict can only be done via the hacky intermediary export to a DEF file) and live with the fact that the associated pdb will embed the previous library name.

This patch just doesn't use the hash for cdylib library crates. It solves the problem above and all tests still pass so I assume it's not breaking anything crucial that was depending on this before 😃

### CODE REVIEW

**Comment by hlopko** [comment]
Looks good, could you please add a test?

File: rust/private/rust.bzl [hunk 1]
```diff
@@ -61,14 +61,14 @@ def _assert_correct_dep_mapping(ctx):
...
-def _determine_lib_name(name, crate_type, toolchain, lib_hash = ""):
+def _determine_lib_name(name, crate_type, toolchain, lib_hash = None):
     """See https://github.com/bazelbuild/rules_rust/issues/405
 
     Args:
         name (str): The name of the current target
         crate_type (str): The `crate_type`
         toolchain (rust_toolchain): The current `rust_toolchain`
-        lib_hash (str, optional): The hashed crate root path. Defaults to "".
+        lib_hash (str, optional): The hashed crate root path. Defaults to None.
```
Comment (UebelAndre) [inline rust/private/rust.bzl:71]: No need to double document the default 😅 (this is my fault for using an auto-docs tool)
```suggestion
        lib_hash (str, optional): The hashed crate root path.
```

**Review by UebelAndre — APPROVED** [review]
Small nit from me but otherwise looks good!

File: test/unit/cdylib_name/cdylib_name_analysis_test.bzl [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    target = analysistest.target_under_test(env)
+
+    file = target[DefaultInfo].files.to_list()[0]
+
+    if file.extension == "so":
+        asserts.equals(env, file.basename, "libsomething.so")
+    if file.extension == "dylib":
+        asserts.equals(env, file.basename, "libsomething.dylib")
+    if file.extension == "dll":
+        asserts.equals(env, file.basename, "something.dll")
```
Comment (UebelAndre) [inline test/unit/cdylib_name/cdylib_name_analysis_test.bzl:17]: These would be mutually exclusive, right?
```python
    if file.extension == "so":
        asserts.equals(env, file.basename, "libsomething.so")
    elif file.extension == "dylib":
        asserts.equals(env, file.basename, "libsomething.dylib")
    elif file.extension == "dll":
        asserts.equals(env, file.basename, "something.dll")
    else:
        analysistest.fail(env, "unexpected file extension for file: {}".format(file.basename))
```

File: test/unit/cdylib_name/cdylib_name_analysis_test.bzl [hunk 3]
```diff
@@ -0,0 +1,42 @@
+"""Analysis tests for the name we assign to cdylib libraries."""
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts")
+load("//rust:defs.bzl", "rust_shared_library")
+
+def _cdylib_name_test_impl(ctx):
+    env = analysistest.begin(ctx)
+    target = analysistest.target_under_test(env)
+
+    file = target[DefaultInfo].files.to_list()[0]
```
Comment (UebelAndre) [inline test/unit/cdylib_name/cdylib_name_analysis_test.bzl:10]:
```suggestion
    # Expecting the `cdylib` to be the one and only file.
    files = target[DefaultInfo].files.to_list()
    asserts.equals(env, len(files), 1)
    file = files[0]
```

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Sorry to flip-flop 😅 I noticed two things before merging

File: test/unit/cdylib_name/cdylib_name_analysis_test.bzl (hunk 3 above)
Comment (ddeville) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl:10]: on windows there will also be an interface library but I can do the appropriate check in each branch.

Comment (UebelAndre) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl]: I'm a big fan of checks that extract what's expected as explicitly as possible. If the files you're after aren't in an `OutputGroupinfo` provider or something, then I think there should be an assert that you can find a unique file from a depset. In general I've found asserting on the number of elements in the depset to cover most edge cases but can be annoying for volatile rules (don't think that's the case here).

Comment (ddeville) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl]: ok I've shuffled things a bit so that we check for the file count and all files in the group are what they are expected.

File: test/unit/cdylib_name/cdylib_name_analysis_test.bzl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    # contain a pair of `.dll` and `.lib` files.
+    files = target[DefaultInfo].files.to_list()
+    if len(files) == 1:
+        asserts.true(env, files[0].extension in ("so", "dylib"))
+        if files[0].extension == "so":
+            asserts.equals(env, files[0].basename, "libsomething.so")
+        elif files[0].extension == "dylib":
+            asserts.equals(env, files[0].basename, "libsomething.dylib")
+    elif len(files) == 2:
+        asserts.equals(env, files[0].basename, "something.dll")
```
Comment (UebelAndre) [inline test/unit/cdylib_name/cdylib_name_analysis_test.bzl:20]: This might be a minor thing but I'd like to avoid creating a dependency on the ordering of the files returned by `DefaultInfo`. I think everyone should do due-diligence to ensure the files they're trying to get out of a list is what they want. Depending on `depset().to_list()[INDEX]` is something I'm hoping we can avoid
```suggestion
        expected_extensions = ["dll", "lib"]
        for file in files:
            asserts.true(env, file.extension in expected_extensions)
            expected_extensions.pop(file.extension)
```

**Review by UebelAndre — CHANGES_REQUESTED** [review]

File: test/unit/cdylib_name/cdylib_name_analysis_test.bzl (hunk 4 above)
Comment (ddeville) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl:20]: oh right that's fair. I'll change it to `expected_filenames` so that we can actually test the cdylib name too in one pass.

Comment (UebelAndre) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl]: Sorry, my understanding of `pop` seems to be wrong. I think I meant [remove](https://docs.bazel.build/versions/4.0.0/skylark/lib/list.html#remove)

Comment (ddeville) [reply test/unit/cdylib_name/cdylib_name_analysis_test.bzl]: oops, fixed (that'll teach me not to test my windows changes on linux, I guess that's why we have CI 😄)

**Review by UebelAndre — APPROVED** [review]
